sugarcrm/client.py

```diff
@@ -0,0 +1,70 @@
+import urllib
+import json
+import hashlib
+import sugarcrm
+from sugarcrm.error import CrmAuthenticationError, CrmResponseError
+
+
+class APIClient:
+    def __init__(self, client=urllib.request, logger=sugarcrm.logger):
+        self.client = client
+        self.session_id = None
+        self.rest_url = sugarcrm.rest_url
+        self.logger = logger
+        encode = hashlib.md5(sugarcrm.password.encode("utf-8"))
+        encodedPassword = encode.hexdigest()
+        args = {"user_auth": {"user_name": sugarcrm.username, "password": encodedPassword}}
+        response = self._api_request("login", args)
+        result = json.loads(response.decode("utf-8"))
+        if "id" in result:
+            self.session_id = result["id"]
+        else:
+            raise CrmAuthenticationError("CRM Invalid Authentication")
+
+    def _api_request(self, method, args):
+        data = json.dumps(args)
+        args = {
+            "method": method,
+            "input_type": "json",
+            "response_type": "json",
+            "rest_data": data,
+        }
+        self.logger.debug("SugarCRM request to {}, args={}".format(self.rest_url, args))
+        params = urllib.parse.urlencode(args).encode("utf-8")
+        response = self.client.urlopen(self.rest_url, params)
+        if not response:
+            raise CrmResponseError("No HTTP Response from {}".format(method))
```

Comment:
Is this even possible? AFAIK all HTTP requests have a response :thinking:

Reply:
Not sure if when I added this I was having some sort of response object that was False in some malfunction sense, but yeah, can't see this happening now and if we change to "requests", for sure not.